std/conv.d

```diff
@@ -4277,1039 +4277,44 @@ private bool isOctalLiteral(const string num)
-        __std_conv_S s;
-        ref __std_conv_S foo() return @property {s.i = j; return s;}
-        alias foo this;
-    }
-}
-
-@system unittest
-{
-    static assert(is(__std_conv_SS : __std_conv_S));
-    __std_conv_S s = void;
-    __std_conv_SS ss = __std_conv_SS(1);
-
-    __std_conv_S sTest1 = ss; //this calls "SS alias this" (and not "S.this(SS)")
-    emplace(&s, ss); //"alias this" should take precedence in emplace over "opCall"
-    assert(s.i == 1);
-}
-
-//Nested classes
-@system unittest
-{
-    class A{}
-    static struct S
-    {
-        A a;
-    }
-    S s1 = void;
-    S s2 = S(new A);
-    emplace(&s1, s2);
-    assert(s1.a is s2.a);
-}
+emplaceRef is a package function for phobos internal use.
+It's implementation has been moved to `core.internal.lifetime`.
++/
+package import core.internal.lifetime : emplaceRef;
 
-//safety & nothrow & CTFE
-@system unittest
+version (StdDdoc)
 {
-    //emplace should be safe for anything with no elaborate opassign
-    static struct S1
-    {
-        int i;
-    }
-    static struct S2
-    {
-        int i;
-        this(int j)@safe nothrow{i = j;}
-    }
-
-    int i;
-    S1 s1 = void;
-    S2 s2 = void;
+    /**
+    Given a raw memory area `chunk`, constructs an object of type `T` in-place.
+    The constructor is passed the arguments `args`, if any.
 
-    auto pi = &i;
-    auto ps1 = &s1;
-    auto ps2 = &s2;
+    For the complete documentation please consult the
+    $(LINK2 https://dlang.org/phobos-prerelease/core/lifetime.html, lifetime) module.
 
-    void foo() @safe nothrow
-    {
-        emplace(pi);
-        emplace(pi, 5);
-        emplace(ps1);
-        emplace(ps1, 5);
-        emplace(ps1, S1.init);
-        emplace(ps2);
-        emplace(ps2, 5);
-        emplace(ps2, S2.init);
-    }
-    foo();
+    Note:
+    This function can be `@trusted` if the corresponding constructor of `T` is `@safe`.
 
-    T bar(T)() @property
-    {
-        T t/+ = void+/; //CTFE void illegal
-        emplace(&t, 5);
-        return t;
-    }
-    // CTFE
-    enum a = bar!int;
-    static assert(a == 5);
-    enum b = bar!S1;
-    static assert(b.i == 5);
-    enum c = bar!S2;
-    static assert(c.i == 5);
-    // runtime
-    auto aa = bar!int;
-    assert(aa == 5);
-    auto bb = bar!S1;
-    assert(bb.i == 5);
-    auto cc = bar!S2;
-    assert(cc.i == 5);
-}
+    Returns: The newly constructed object.
+     */
+    T* emplace(T)(T* chunk) @safe pure nothrow;
 
+    /// ditto
+    T* emplace(T, Args...)(T* chunk, auto ref Args args);
 
-@system unittest
-{
-    struct S
-    {
-        int[2] get(){return [1, 2];}
-        alias get this;
-    }
-    struct SS
-    {
-        int[2] ii;
-    }
-    struct ISS
-    {
-        int[2] ii;
-    }
-    S s;
-    SS ss = void;
-    ISS iss = void;
-    emplace(&ss, s);
-    emplace(&iss, s);
-    assert(ss.ii == [1, 2]);
-    assert(iss.ii == [1, 2]);
-}
+    /// ditto
+    T emplace(T, Args...)(T chunk, auto ref Args args);
```

Comment:
Don't forget to include the template constraints for the docs - e.g. `if (is(T == class))`, IIRC, for this overload.

Reply:
I don't think I can add the template constraint in the `function declaration`. I tried to do that, but I couldn't get it to compile.

Reply:
Since this will only exist when building the docs, you can use a full function definition with a trivial body - e.g. `{ return chunk; }`

Reply:
@ZombineDev added the template constraints for the docs.